nixos/modules/services/home-automation/esphome.nix

```diff
@@ -0,0 +1,153 @@
+      description = mdDoc "Whether to open the firewall for the specified port.";
+    };
+
+    allowedDevices = mkOption {
+      default = [];
+      example = [{ node = "/dev/serial/by-id/usb-Silicon_Labs_CP2102_USB_to_UART_Bridge_Controller_0001-if00-port0"; modifier = "rw"; }];
+      description = lib.mdDoc ''
+        A list of device nodes to which {command}`esphome` has access to.
+        Beware that permissions are not added dynamically when a device
+        is plugged in while the service is already running.
+      '';
+      type = types.listOf (types.submodule {
+        options = {
+          node = mkOption {
+            example = "/dev/ttyUSB*";
+            type = types.str;
+            description = lib.mdDoc "Path to device node";
+          };
+          modifier = mkOption {
+            example = "rw";
+            type = types.str;
+            description = lib.mdDoc ''
+              Device node access modifier. Takes a combination
+              `r` (read), `w` (write), and `m` (mknod). See the
+              `systemd.resource-control(5)` man page for more
+              information.
+            '';
+          };
+        };
+      });
+    };
+  };
+
+  config = mkIf cfg.enable {
+    networking.firewall.allowedTCPPorts = mkIf (cfg.openFirewall && !cfg.enableUnixSocket) [cfg.port];
+
+    systemd.services.esphome = {
+      description = "ESPHome dashboard";
+      after = ["network.target"];
+      wantedBy = ["multi-user.target"];
+      path = [cfg.package];
+
+      # platformio fails to determine the home directory when using DynamicUser
+      environment.PLATFORMIO_CORE_DIR = "${stateDir}/.platformio";
+
+      serviceConfig = {
+        ExecStart = "${cfg.package}/bin/esphome dashboard ${esphomeParams} ${stateDir}";
+        DynamicUser = true;
+        User = "esphome";
+        Group = "esphome";
+        WorkingDirectory = stateDir;
+        StateDirectory = "esphome";
+        StateDirectoryMode = "0750";
+        Restart = "on-failure";
+        RuntimeDirectory = mkIf cfg.enableUnixSocket "esphome";
+        RuntimeDirectoryMode = "0750";
+
+        # Hardening
+        CapabilityBoundingSet = "";
+        LockPersonality = true;
+        MemoryDenyWriteExecute = true;
+        DevicePolicy = "closed";
+        DeviceAllow = map (d: "${d.node} ${d.modifier}") cfg.allowedDevices;
+        SupplementaryGroups = ["dialout"];
+        NoNewPrivileges = true;
+        PrivateUsers = true;
+        PrivateTmp = true;
+        ProtectClock = true;
+        ProtectControlGroups = true;
+        ProtectHome = true;
+        ProtectHostname = true;
+        ProtectKernelLogs = true;
+        ProtectKernelModules = true;
+        ProtectKernelTunables = true;
+        ProtectProc = "invisible";
+        ProcSubset = "pid";
+        ProtectSystem = "strict";
+        RemoveIPC = true;
+        RestrictAddressFamilies = [
+          "AF_INET"
+          "AF_INET6"
+          "AF_NETLINK"
+          "AF_UNIX"
```

Comment:
First Question: Can we only allow the `AF_INET` and `AF_INET6` when the tcp variant is being used and only `AF_UNIX` when the unix socket is enabled?
Second Question: Is the `AF_NETLINK` really needed, if so , for that is it being used ?

Reply:
Also one more note for the `AF_INET6`, maybe we want to tie it to the enablement of `networking.enableIPv6` https://search.nixos.org/options?channel=unstable&show=networking.enableIPv6&from=0&size=50&sort=relevance&type=packages&query=networking.enableIPv6 but that may happen automatically already, I dont know.

Reply:
I already tried restricting this further, but removing any of these results in the application not working. Removing `AF_NETLINK` results in `OSError: [Errno 97] Address family not supported by protocol`. My only guess is that this is related to using `python-zeroconf`.